packages/client/src/fetchMetadata.ts

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2024 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type {
+  ActionDefinition,
+  InterfaceDefinition,
+  ObjectTypeDefinition,
+  QueryDefinition,
+} from "@osdk/api";
+import type {
+  ActionMetadata,
+  Definition,
+  InterfaceMetadata,
+  ObjectMetadata,
+  QueryMetadata,
+} from "@osdk/client.api";
+import { OntologiesV2 } from "@osdk/internal.foundry";
+import type { MinimalClient } from "./MinimalClientContext.js";
+import { addUserAgentAndRequestContextHeaders } from "./util/addUserAgentAndRequestContextHeaders.js";
+
+/** @internal */
+export const fetchMetadataInternal = async <
+  Q extends Definition,
+>(
+  client: MinimalClient,
+  definition: Q,
+): Promise<
+  Q extends ObjectTypeDefinition<any, any> ? ObjectMetadata
+    : Q extends InterfaceDefinition<any, any> ? InterfaceMetadata
+    : Q extends ActionDefinition<any, any, any> ? ActionMetadata
+    : Q extends QueryDefinition<any, any, any> ? QueryMetadata
+    : never
+> => {
+  if (definition.type === "object") {
+    return fetchObjectMetadata(client, definition) as any;
+  } else if (definition.type === "interface") {
+    return fetchInterfaceMetadata(client, definition) as any;
+  } else if (definition.type === "action") {
+    return fetchActionMetadata(client, definition) as any;
+  } else {
```

Comment:
I have to use else instead of else if here even though this is the only other option because otherwise I get ts errors saying I cannot return undefined. Is there a clever way to get around this?

Reply:
Do the `else if { ... } else { throw "Should never happen": }` ?

Reply:
Ah yeah that will work, but is there not a way to avoid throwing an error and have it fully checked by the compiler

Reply:
Never mind, I was just a bit confused, was basically looking for an assertNever here